src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+        private bool DeconstructWithDeconstruct(
+                        BoundDeconstructValuePlaceholder targetPlaceholder,
+                        AssignmentExpressionSyntax syntax,
+                        DiagnosticBag diagnostics,
+                        ImmutableArray<BoundExpression> variables,
+                        ArrayBuilder<BoundDeconstructionAssignmentOperator.DeconstructStep> deconstructionSteps,
+                        ArrayBuilder<BoundDeconstructionAssignmentOperator.AssignmentStep> assignmentSteps)
         {
             // symbol and parameters for Deconstruct
             DiagnosticBag bag = new DiagnosticBag();
-            MethodSymbol deconstructMethod = FindDeconstruct(checkedVariables, boundRHS, node, bag);
+            MethodSymbol deconstructMethod = FindDeconstruct(variables.Length, targetPlaceholder, syntax, bag);
             if (!diagnostics.HasAnyErrors())
             {
                 diagnostics.AddRange(bag);
             }
 
             if ((object)deconstructMethod == null)
             {
-                return new BoundDeconstructionAssignmentOperator(node, checkedVariables, boundRHS, ErrorMethodSymbol.UnknownMethod,
-                                                                ImmutableArray<BoundDeconstructionAssignmentOperator.AssignmentInfo>.Empty,
-                                                                ErrorTypeSymbol.UnknownResultType,
-                                                                hasErrors: true);
+                return false;
             }
 
-            // figure out the pairwise conversions
             var deconstructParameters = deconstructMethod.Parameters;
-            var assignments = checkedVariables.SelectAsArray((variable, index, parameters) => MakeAssignmentInfo(variable, parameters[index].Type, node, diagnostics), deconstructParameters);
 
-            TypeSymbol voidType = GetSpecialType(SpecialType.System_Void, diagnostics, node);
-            return new BoundDeconstructionAssignmentOperator(node, checkedVariables, boundRHS, deconstructMethod, assignments, voidType);
+            // add step for deconstructing this target
+            var deconstructionStep = new BoundDeconstructionAssignmentOperator.DeconstructStep()
+            {
+                DeconstructMemberOpt = deconstructMethod,
+                TargetPlaceholder = targetPlaceholder,
+                OutputPlaceholders = deconstructParameters.SelectAsArray(parameter => new BoundDeconstructValuePlaceholder(syntax, parameter.Type))
+            };
+            deconstructionSteps.Add(deconstructionStep);
+
+            // outputs will either need a conversion step and assignment step, or if they are nested variables, they will need further deconstruction
+            if (!DeconstructOrAssignOutputs(deconstructionStep, variables, syntax, diagnostics, deconstructionSteps, assignmentSteps))
+            {
+                return false;
+            }
+
+            return true;
```

Comment:
Yes, I can indeed push the code factoring farther. Thanks!
I refactored as you suggested, but changed the `DeconstructionStep` type back to a class (so that I can use null for error cases).